core/src/main/scala/org/apache/spark/shuffle/sort/SortShuffleManager.scala

```diff
@@ -75,9 +77,10 @@ private[spark] class SortShuffleManager(conf: SparkConf) extends ShuffleManager
   }
 
   /**
-   * A mapping from shuffle ids to the number of mappers producing output for those shuffles.
+   * A mapping from shuffle ids to the tuple of number of mappers producing output and
+   * indeterminate stage attempt id for those shuffles.
    */
-  private[this] val numMapsForShuffle = new ConcurrentHashMap[Int, Int]()
+  private[this] val infoMapsForShuffle = new ConcurrentHashMap[Int, (Int, Option[Int])]()
```

Comment:
It's possible that there are 2 running map tasks with different shuffle generation id. What will happen with your change?

Reply:
Thanks for reminding, I'll try to create a case of 2 running map stage based on the description in  https://issues.apache.org/jira/browse/SPARK-7308 and check the behavior.